### PR TITLE
feat: Allow DataFrame.join for self-join on Null index

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

**Review by tswast — COMMENTED** [review]
Thanks Trevor!

Should we add an e2e test for the .fit(X, y) table with null index / unordered mode, too?

**Comment by TrevorBergeron** [comment]
> Thanks Trevor!
> 
> Should we add an e2e test for the .fit(X, y) table with null index / unordered mode, too?

Good idea. This revealed that ml modules were caching pre-join, which is invalidated by row-identity join. Instead, I made it cache post-join.

File: bigframes/ml/core.py [hunk 1]
```diff
@@ -326,7 +326,7 @@ def create_model(
         if y_train is None:
             input_data = X_train.cache()
         else:
-            input_data = X_train.cache().join(y_train.cache(), how="outer")
+            input_data = X_train.join(y_train.cache(), how="outer").cache()
```
Comment (GarrettWu) [inline bigframes/ml/core.py:329]: this should be y_train without cache() as well?

File: tests/system/large/ml/test_linear_model.py [hunk 2]
```diff
@@ -150,6 +150,44 @@ def test_logistic_regression_configure_fit_score(penguins_df_default_index, data
     assert reloaded_model.class_weight is None
 
 
+def test_unordered_mode_logistic_regression_configure_fit_score(
+    unordered_session, penguins_table_id, dataset_id
+):
+    model = bigframes.ml.linear_model.LogisticRegression()
```
Comment (GarrettWu) [inline tests/system/large/ml/test_linear_model.py:156]: Not a problem, but if we only pick one model to test some shared functionalities, usually the choice is LinearReg.

Comment (TrevorBergeron) [reply tests/system/large/ml/test_linear_model.py]: ok, used linear regression model instead

File: bigframes/ml/core.py (hunk 1 above)
Comment (TrevorBergeron) [reply bigframes/ml/core.py]: yes, fixed

**Review by tswast — APPROVED** [review]

**Pull request merged**